transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java

```diff
@@ -311,19 +317,24 @@ private void invokeChannelRead(Object msg) {
         }
     }
 
+    private void invokeNextChannelRead(AbstractChannelHandlerContext next, Object msg) {
+        nextChannelReadInvoked = true;
+        next.invokeChannelRead(msg);
+    }
+
     @Override
     public ChannelHandlerContext fireChannelReadComplete() {
         final AbstractChannelHandlerContext next = findContextInbound();
         EventExecutor executor = next.executor();
         if (executor.inEventLoop()) {
-            next.invokeChannelReadComplete();
+            invokeNextChannelReadComplete(next);
         } else {
             Runnable task = next.invokeChannelReadCompleteTask;
             if (task == null) {
                 next.invokeChannelReadCompleteTask = task = new Runnable() {
                     @Override
                     public void run() {
-                        next.invokeChannelReadComplete();
+                        invokeNextChannelReadComplete(next);
```

Comment:
This is actually not a good idea because the next context will then have a reference to the current context. i.e. we replaced next.invoke*() to this.invokeNext*(). I think this is potentially very difficult to follow?